api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java

```diff
@@ -0,0 +1,190 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.joda.time.DateTime;
+
+import com.google.gson.Gson;
+import com.takipi.api.client.ApiClient;
+import com.takipi.api.client.result.event.EventResult;
+import com.takipi.api.client.util.event.EventUtil;
+import com.takipi.api.client.util.regression.RateRegression;
+import com.takipi.api.client.util.regression.RegressionInput;
+import com.takipi.api.client.util.regression.RegressionStringUtil;
+import com.takipi.api.client.util.regression.RegressionUtil;
+
+public class ProcessQualityGates {
+
+	private static QualityGateReport qualityReport;
+	
+	//process CICD gates based on the inputs sent in
+	public static QualityGateReport processCICDInputs(ApiClient apiClient, RegressionInput input, boolean newEvents,
+			boolean resurfacedEvents, String regexFilter, int topIssuesVolume, boolean countGate, PrintStream printStream, boolean verbose) {
+		
+		qualityReport = new QualityGateReport();
+		
+		DateTime deploymentStart = RegressionUtil.getDeploymentStartTime(apiClient, input.serviceId, input.deployments);
+		
+		if (deploymentStart == null) {
+			throw new IllegalStateException("Deployment name " + input.deployments + " not found. Please ensure your collector and Jenkins configuration are pointing to the same enviornment.");
+		}
+		
+		Collection<EventResult> events = RegressionUtil.getActiveEventVolume(apiClient, input, deploymentStart,
+				printStream);
+		
+		if (events == null) {
+			events = new ArrayList<EventResult>();
+		}
+		
+		//filter out unneeded events set in plugin (regexFilter)
+		events = filterEvents(events, regexFilter);
+		
+		//check count gates
+		if (countGate) {
+			events = RateRegression.getSortedNewEvents(events);
+			qualityReport.setTopErrors(getTopXEvents(apiClient, input, events, topIssuesVolume, deploymentStart));
+		}
+		
+		//find new Errors
+		if (newEvents) {
+			qualityReport.setNewErrors(getNewErrors(apiClient, input, events, deploymentStart));
+		}
+		
+		//find resurfaced Errors
+		if (resurfacedEvents) {
+			qualityReport.setResurfacedErrors(getResurfacedErrors(apiClient, input, events, deploymentStart));
+		}
+		
+		//find critical errors
+		if (input.criticalExceptionTypes != null) {
+			qualityReport.setCriticalErrors(getCriticalErrors(apiClient, input, events, deploymentStart));
+		}
+		
+		return qualityReport;
+	}
+	
+	private static List<OOReportEvent> getTopXEvents(ApiClient apiClient, RegressionInput input,
+			Collection<EventResult> events, int topIssuesVolume, DateTime deploymentStart) {
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();
+		
+		int counter = 1;
+		for (EventResult event : events) {
+			String arcLink = getArcLink(apiClient, event.id, input, deploymentStart);
+			OOReportEvent newEvent = new OOReportEvent(event, null, arcLink);
+			result.add(newEvent);
+			
+			if (counter == topIssuesVolume) {
```

Comment:
Can check `topIssuesVolume` directly against `result.size()`
(The `size` method for `ArrayList` runs at constant time, and is optimized by the JVM so no performance overhead for calling it a lot)

Reply:
Nice. Changed